File: src/main/java/tk/hiddenname/smarthome/service/hardware/manager/DeviceManager.java
package tk.hiddenname.smarthome.service.hardware.manager;

import lombok.AllArgsConstructor;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;
import tk.hiddenname.smarthome.exception.GPIOBusyException;
import tk.hiddenname.smarthome.exception.PinSignalSupportException;
import tk.hiddenname.smarthome.exception.SignalTypeNotFoundException;
import tk.hiddenname.smarthome.model.hardware.Device;
import tk.hiddenname.smarthome.model.hardware.GPIO;
import tk.hiddenname.smarthome.model.signal.SignalType;
import tk.hiddenname.smarthome.service.database.DeviceDatabaseService;
import tk.hiddenname.smarthome.service.hardware.impl.GPIOService;
import tk.hiddenname.smarthome.service.hardware.impl.digital.output.DigitalDeviceServiceImpl;
import tk.hiddenname.smarthome.service.hardware.impl.pwm.output.PwmDeviceServiceImpl;

import java.util.Objects;

@Component
@AllArgsConstructor
public class DeviceManager {

    private static final Logger log = LoggerFactory.getLogger(DeviceManager.class);

    private final DigitalDeviceServiceImpl digitalService;
    private final PwmDeviceServiceImpl pwmService;
    private final DeviceDatabaseService service;

    public void create(Device device) throws PinSignalSupportException, GPIOBusyException {
        log.debug("Creating device " + device.toString());
        GPIO gpio = device.getGpio();
        assert gpio != null;
        getService(Objects.requireNonNull(gpio.getType())).save(
                device.getId(),
                gpio.getGpioPin(),
                device.getSignalInversion()
        );
    }

    public void update(Device device) {
        log.debug("Updating device " + device.toString());
        GPIO gpio = device.getGpio();
        assert gpio != null;
        getService(Objects.requireNonNull(gpio.getType())).update(
                device.getId(),
                device.getSignalInversion()
        );
    }

    public void delete(Device device) {
        log.debug("Deleting device " + device.toString());
        GPIO gpio = device.getGpio();
        assert gpio != null;
        getService(Objects.requireNonNull(gpio.getType())).delete(
                device.getId()
        );
    }

    public void loadDevices() {
        for (Device device : service.getAll()) {
            try {
                create(device);
            } catch (PinSignalSupportException | GPIOBusyException e) {
                log.warn(e.getMessage());
            }
        }
    }

    private GPIOService getService(SignalType type) {
        switch (type) {
            case DIGITAL:
                return digitalService;
            case PWM:
                return pwmService;
            default:
                SignalTypeNotFoundException e = new SignalTypeNotFoundException(type.toString());
                log.warn(e.getMessage());
                throw e;
        }
    }
}

File: src/main/kotlin/tk/hiddenname/smarthome/service/hardware/manager/DeviceManager.kt
package tk.hiddenname.smarthome.service.hardware.manager

import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component
import tk.hiddenname.smarthome.exception.exist.GpioPinBusyException
import tk.hiddenname.smarthome.exception.not_specified.GpioNotSpecifiedException
import tk.hiddenname.smarthome.exception.not_specified.SignalTypeNotSpecifiedException
import tk.hiddenname.smarthome.exception.support.PinSignalSupportException
import tk.hiddenname.smarthome.exception.support.UnsupportedSignalTypeException
import tk.hiddenname.smarthome.model.hardware.Device
import tk.hiddenname.smarthome.model.signal.SignalType
import tk.hiddenname.smarthome.service.database.DeviceDatabaseService
import tk.hiddenname.smarthome.service.hardware.impl.GPIOService
import tk.hiddenname.smarthome.service.hardware.impl.digital.output.DigitalDeviceService
import tk.hiddenname.smarthome.service.hardware.impl.pwm.output.PwmDeviceServiceImpl
import tk.hiddenname.smarthome.utils.gpio.GpioManager

@Component
class DeviceManager(
    private val digitalService: DigitalDeviceService,
    private val pwmService: PwmDeviceServiceImpl,
    private val service: DeviceDatabaseService,
    private val gpioManager: GpioManager
) {

    private val log = LoggerFactory.getLogger(DeviceManager::class.java)

    @Throws(PinSignalSupportException::class, GpioPinBusyException::class)
    fun register(device: Device) {
        log.info("Creating device $device")
        device.gpio ?: throw GpioNotSpecifiedException()
        getService(device.gpio.signalType).save(
            device.id,
            device.gpio,
            device.signalInversion
        )
    }

    fun update(device: Device) {
        log.info("Updating device $device")
        device.gpio ?: throw GpioNotSpecifiedException()
        getService(device.gpio.signalType).update(
            device.id,
            device.signalInversion
        )
    }

    fun unregister(device: Device) {
        log.info("Deleting device $device")
        device.gpio ?: throw GpioNotSpecifiedException()
        getService(device.gpio.signalType).delete(
            device.id
        )
    }

    fun loadDevices() {
        service.getAll().forEach {
            try {
                register(it)
            } catch (e: PinSignalSupportException) {
                log.warn(e.message)
            }
        }
        log.debug("Devices loaded")
    }

    @Suppress("DuplicatedCode")
    fun changeSignalType(device: Device, newSignalType: SignalType?) {
        newSignalType ?: throw SignalTypeNotSpecifiedException()
        device.gpio ?: throw GpioNotSpecifiedException()
        if (gpioManager.validate(device.gpio.gpioPin, newSignalType, device.gpio.pinMode)) {
            unregister(device)
            device.gpio.signalType = newSignalType
            register(device)
        }
    }

    private fun getService(type: SignalType?): GPIOService {
        return when (type) {
            SignalType.DIGITAL -> digitalService
            SignalType.PWM -> pwmService
            else -> throw UnsupportedSignalTypeException(type!!)
        }
    }
}